src/mds/ScrubStack.h

```diff
@@ -55,88 +55,77 @@ class ScrubStack {
   MDCache *mdcache;
   ScrubStack(MDCache *mdc, Finisher *finisher_) :
     finisher(finisher_),
-    dentry_stack(member_offset(CDentry, item_scrub)),
+    inode_stack(member_offset(CInode, item_scrub)),
     scrubs_in_progress(0),
     scrubstack(this),
     stack_size(0),
     scrub_kick(mdc, this),
     mdcache(mdc) {}
   ~ScrubStack() {
-    assert(dentry_stack.empty());
+    assert(inode_stack.empty());
     assert(!scrubs_in_progress);
   }
   /**
-   * Put a dentry on the top of the scrub stack, so it is the highest priority.
+   * Put a inode on the top of the scrub stack, so it is the highest priority.
    * If there are other scrubs in progress, they will not continue scrubbing new
    * entries until this one is completed.
-   * @param dn The dentry to scrub
-   * @param recursive True if we want to recursively scrub the
-   * entire hierarchy under dn.
-   * @param children True if we want to scrub the direct children of
-   * dn but aren't doing a recursive scrub. (Otherwise, all checks are
-   * local to dn's disk state.)
+   * @param in The inodey to scrub
    * @param header The ScrubHeader propagated from whereever this scrub
    *               was initiated
    */
-  void enqueue_dentry_top(CDentry *dn, bool recursive, bool children,
-                          ScrubHeaderRefConst header,
-                          MDSInternalContextBase *on_finish) {
-    enqueue_dentry(dn, recursive, children, header, on_finish, true);
+  void enqueue_inode_top(CInode *in, const ScrubHeaderRefConst& header,
+			 Context *on_finish) {
```

Comment:
These changes to take a Context instead of an MDSInternalContextBase are losing us some constraints on the locking associated with those parameters. Is there some change making them safe? It looks like you made the change to accommodate C_MDS_EnqueueScrub, shouldn't the callbacks for admin sockets have a layer between our internal ScrubStack interfaces and reactivating them directly?